aptos-move/framework/src/natives/aggregator_natives/helpers_v2.rs

```diff
@@ -1,48 +1,144 @@
-
-pub(crate) fn aggregator_value_field_as_id(
-    value: u128,
-    resolver: &dyn DelayedFieldResolver,
-) -> PartialVMResult<DelayedFieldID> {
-    let value_u64 = u128_to_u64(value)?;
-    Ok(resolver.validate_and_convert_delayed_field_id(value_u64)?)
+use crate::natives::aggregator_natives::aggregator_v2::{
+    EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE, EUNSUPPORTED_AGGREGATOR_TYPE,
+};
+use aptos_aggregator::resolver::DelayedFieldResolver;
+use aptos_native_interface::{safely_get_struct_field_as, SafeNativeError, SafeNativeResult};
+use aptos_types::delayed_fields::{string_to_bytes, DelayedFieldID};
+use move_binary_format::errors::PartialVMError;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Reference, SizedID, Struct, StructRef, Value},
+};
+
+// Field indices for aggregator Move struct.
+const AGGREGATOR_VALUE_FIELD_INDEX: usize = 0;
+const AGGREGATOR_MAX_VALUE_FIELD_INDEX: usize = 1;
+
+// Field indices for aggregator snapshot Move struct.
+const AGGREGATOR_SNAPSHOT_VALUE_FIELD_INDEX: usize = 0;
+
+// Field indices for derived string snapshot Move struct.
+const DERIVED_STRING_SNAPSHOT_VALUE_FIELD_INDEX: usize = 0;
+const _DERIVED_STRING_SNAPSHOT_PADDING_FIELD_INDEX: usize = 1;
+
+macro_rules! get_value_impl {
+    ($func_name:ident, $idx:expr, $e:expr) => {
+        pub(crate) fn $func_name(struct_ref: &StructRef, ty: &Type) -> SafeNativeResult<u128> {
+            Ok(match ty {
+                Type::U128 => safely_get_struct_field_as!(struct_ref, $idx, u128),
+                Type::U64 => safely_get_struct_field_as!(struct_ref, $idx, u64) as u128,
+                _ => return Err(SafeNativeError::Abort { abort_code: $e }),
+            })
+        }
+    };
 }
 
-pub(crate) fn aggregator_snapshot_value_field_as_id(
-    value: u128,
-    resolver: &dyn DelayedFieldResolver,
-) -> PartialVMResult<DelayedFieldID> {
-    aggregator_value_field_as_id(value, resolver)
+macro_rules! get_value_as_id_impl {
+    ($func_name:ident, $idx:expr, $e:expr) => {
+        pub(crate) fn $func_name(
+            struct_ref: &StructRef,
+            ty: &Type,
+            resolver: &dyn DelayedFieldResolver,
+        ) -> SafeNativeResult<DelayedFieldID> {
+            let id = match ty {
+                Type::U64 | Type::U128 => {
+                    let sized_id =
+                        safely_get_struct_field_as!(struct_ref, $idx, SizedID);
+                    sized_id.into()
+                },
+                _ => {
+                    return Err(SafeNativeError::Abort {
+                        abort_code: $e,
+                    })
+                },
+            };
+
+            // Make sure we validate generated id is correct, i.e. it lies within the
+            // right bounds, etc. to catch bugs early.
+            resolver
+                .validate_delayed_field_id(&id)
+                .map_err(|e| SafeNativeError::InvariantViolation(PartialVMError::from(e)))?;
```

Comment:
since VM is the one that created those ValueImpls, this should never happen without code bugs - and so should be (delayed fields) code invariant error instead

Reply:
We are inside a native function, so as far as I understand the way it works is that you would match on SafeNativeError and if it is an invariant violation the error is forwarded:
```
Err(err) => match err {
  Abort { abort_code } => Ok(NativeResult::err(context.gas_used, abort_code)),
  OutOfGas => Ok(NativeResult::out_of_gas(context.gas_used)),
  // TODO(Gas): Check if err is indeed an invariant violation.
  InvariantViolation(err) => Err(err),
},
```
from here: https://github.com/aptos-labs/aptos-core/blob/183c6c8339ae2a6009228d1f1f9abd55aca1b947/aptos-move/aptos-native-interface/src/builder.rs#L139